src/utils/matching.ts

```diff
@@ -0,0 +1,119 @@
+const regexpCache = new Map();
+
+interface MatchInterface {
+    regexp: RegExp;
+    negated: boolean;
+}
+
+function makeRegexp(pattern: string): MatchInterface {
+    if (regexpCache.has(pattern)) {
+        return regexpCache.get(pattern);
+    }
+
+    const negated = pattern[0] === '!';
+    if (negated) {
+        pattern = pattern.substr(1);
+    }
+
+    // Check if the pattern is regex
+    if (pattern[0] === '/') {
+        // Get the flag of the specified regex
+        const flag = pattern.substr(pattern.lastIndexOf('/') + 1);
+        // Remove the / indentifiers so Regexp can make valid regexp of it.
+        pattern = pattern.substr(1).substr(0, pattern.lastIndexOf('/') - 1);
+        const regexp = new RegExp(pattern, flag);
+        const regObject: MatchInterface = {
+            regexp,
+            negated,
+        };
+        return regObject;
+    }
+    // Magic replacement to ensure pattern is valid
+    pattern = pattern.replace(/[|\\{}()[\]^$+*?.]/g, '\\$&').replace(/-/g, '\\x2d').replace(/\\\*/g, '[\\s\\S]*');
+    // Create an "hard" regexp to ensure that it's the exact domain we are matching and not some subdomain.
+    const regexp = new RegExp(`^${pattern}(?![A-Za-z0-9.])`, 'i');
+    const regObject: MatchInterface = {
+        regexp,
+        negated,
+    };
+    // Make sure it's cached!
+    regexpCache.set(pattern, regObject);
+    return regObject;
+}
+
+/**
+ * Sanitazed the website so it ensures different input
+ * with the same meaning will have the same output here.
+ */
+function sanitazeInput(input: string) {
```

Comment:
I'm not sure we should perform this kind of magic transformation. It seems, every time someone does this kind of cleverness something else gets broken in mysterious ways. I remember Google got a lot of bad publicity from _cosmetically_ hiding what they called "trivial subdomains".

Reply:
I'm not sure how adding any of these characters could be "clever" I'm removing `$` and `^` as they don't have much effect anymore. `www.` is to account for old web pages which still uses `www.`, And it removes any query(We just don't want to mess around with it) and we remove the scheme.

Reply:
I'm mostly worried about `www.` subdomain.

Reply:
I can see why someone would do it, but in such rare rare cases they can go for the regex solution.

Reply:
Let's remove the `www.` then.